pkg/controller/jobcontrol.go

```diff
@@ -0,0 +1,271 @@
+		return nil, false, nil
+	}
+
+	// Look through jobs that belong to the owner and that match the type
+	// of job created by this control.
+	// If the job's generation corresponds to the owner's current
+	// generation, sync the owner status with the job's latest status.
+	// If the job does not correspond to the owner's current generation,
+	// delete the job.
+	oldJobs := []*kbatch.Job{}
+	var currentJob *kbatch.Job
+	jobs, err := c.jobsLister.Jobs(owner.GetNamespace()).List(labels.Everything())
+	if err != nil {
+		return nil, false, err
+	}
+	for _, job := range jobs {
+		if !metav1.IsControlledBy(job, owner) {
+			continue
+		}
+		if !c.IsControlledJob(job) {
+			continue
+		}
+		if jobOwnerGeneration(job) == owner.GetGeneration() {
+			currentJob = job
+		} else {
+			logger.WithField("job", jobKey(job)).
+				Debugln("found job that does not correspond to owner's current generation")
+			oldJobs = append(oldJobs, job)
+		}
+	}
+
+	if len(oldJobs) > 0 {
+		return nil, false, c.deleteOldJobs(ownerKey, oldJobs, logger)
+	}
+
+	if currentJob != nil {
+		return currentJob, false, nil
+	}
+
+	if jobFactory != nil {
+		job, err := c.createJob(ownerKey, owner, jobFactory, logger)
+		return job, job != nil, err
+	}
+
+	return nil, false, nil
+}
+
+func (c *jobControl) ObserveOwnerDeletion(ownerKey string) {
+	c.expectations.DeleteExpectations(ownerKey)
+}
+
+func (c *jobControl) ObserveJobCreation(ownerKey string, job *kbatch.Job) {
+	c.expectations.CreationObserved(ownerKey)
+}
+
+func (c *jobControl) ObserveJobDeletion(ownerKey string, job *kbatch.Job) {
+	c.expectations.DeletionObserved(ownerKey, jobKey(job))
+}
+
+func (c *jobControl) IsControlledJob(job *kbatch.Job) bool {
+	return strings.HasPrefix(job.Name, c.jobPrefix)
+}
+
+func (c *jobControl) createJob(ownerKey string, owner metav1.Object, jobFactory JobFactory, logger log.FieldLogger) (*kbatch.Job, error) {
+	logger.Infof("Creating new %q job", c.jobPrefix)
+
+	name := names.SimpleNameGenerator.GenerateName(fmt.Sprintf("%s%s-", c.jobPrefix, owner.GetName()))
+
+	job, configMap, err := jobFactory.BuildJob(name)
+	if err != nil {
+		return nil, err
+	}
+
+	if job == nil {
+		return nil, nil
+	}
+
+	ownerRef := metav1.NewControllerRef(owner, c.ownerKind)
+
+	job.OwnerReferences = append(job.OwnerReferences, *ownerRef)
+	if job.Annotations == nil {
+		job.Annotations = map[string]string{}
+	}
+	job.Annotations[clusteroperator.OwnerGenerationAnnotation] = fmt.Sprintf("%d", owner.GetGeneration())
+
+	cleanUpConfigMap := true
+	if configMap != nil {
+		configMap.OwnerReferences = append(configMap.OwnerReferences, *ownerRef)
+
+		configMap, err = c.kubeClient.CoreV1().ConfigMaps(owner.GetNamespace()).Create(configMap)
+		if err != nil {
+			return nil, err
+		}
+		defer func() {
+			if configMap != nil && cleanUpConfigMap {
+				if err := c.kubeClient.CoreV1().ConfigMaps(configMap.Namespace).Delete(configMap.Name, &metav1.DeleteOptions{}); err != nil {
```

Comment:
We should make an effort to delete the configmap with the same name when we delete the job. I decided not to do that as part of this PR to limit the new functionality introduced by this PR. I will add an issue to make sure that this is tracked for future work.

I don't think that we need to *watch* configmaps. The point of watching job deletes is so that we know when it is safe to start a new job. We don't need to block anything else due to the existence of a configmap.

Note that this delete here is only to clean up the configmap in the case that there was an error that prevented the successful creation of the job.

Reply:
ic, thx that's fine